project/VersionUtil.scala

```diff
@@ -85,12 +86,27 @@ object VersionUtil {
     val date = executeTool("get-scala-commit-date")
     val sha = executeTool("get-scala-commit-sha").substring(0, 7)
 
+    val Patch = """\d+\.\d+\.(\d+)""".r
+    def cross = base match {
+      case Patch(p) if p.toInt > 0 => "bin"
+      case _ => "pre"
+    }
+
     val (canonicalV, mavenSuffix, osgiV, release) = suffix match {
-      case "SNAPSHOT"     => (s"$base-$date-$sha",   s"-SNAPSHOT",      s"$base.v$date-$sha",         false)
-      case "SHA-SNAPSHOT" => (s"$base-$date-$sha",   s"-$sha-SNAPSHOT", s"$base.v$date-$sha",         false)
-      case "SHA-NIGHTLY"  => (s"$base-$sha-nightly", s"-$sha-nightly",  s"$base.v$date-NIGHTLY-$sha", true)
-      case ""             => (s"$base",              "",                s"$base.v$date-VFINAL-$sha",  true)
-      case suffix         => (s"$base-$suffix",      s"-$suffix",       s"$base.v$date-$suffix-$sha", true)
+      // TODO: should this remain `-SNAPSHOT`? does `-SNAPSHOT` have a special meaning for maven / sbt? does `bin-SNAPSHOT` affect this?
+      // TODO: include `-bin` / `-pre` in the canonical version? in the osgi version?
```

Comment:
You're right. I'm blaming lack of vertical alignment making it harder to comprehend.

I see what you're doing and it looks good.